tests/console/snapper_cleanup.pm

```diff
@@ -73,17 +73,53 @@ sub run {
         assert_script_run "snapper set-config NUMBER_LIMIT=2-10 NUMBER_LIMIT_IMPORTANT=4-10 SPACE_LIMIT=0.5";
     }
 
+    my ($n_scratch_prepost, $scratchfile_mb,          $safety_margin_mb,   $initially_free);
+    my ($nmax_snapshots,    $number_limit_upper_orig, $number_limit_upper, $n);
+
+    # three test parameters; hardcoded values found appropriate in past tests
+    $n_scratch_prepost = 3;
+    $scratchfile_mb    = 1024;
+    # Expected to keep the test from filling the last 300 MB of the filesystem, poo#36838
+    $safety_margin_mb = 300 + $scratchfile_mb;
+
     assert_script_run("snapper get-config");    # get initial cfg
+    $number_limit_upper_orig = script_output("snapper get-config | awk '/^NUMBER_LIMIT\\>/{ print \$3 }'");
     assert_script_run("snapper list");          # get initial list of snap's
-    assert_script_run("snapper set-config NUMBER_MIN_AGE=0");
+
+    # check amount of free fs disk space and adapt to it
+    $initially_free = get_space("$btrfs_fs_usage | awk -F ' ' '/Free .estimated.:.*min:/{print\$3}'");    # bytes
+    $nmax_snapshots = int(($initially_free / 1_048_576 - $safety_margin_mb) / $scratchfile_mb) * 2;
+
+    # NUMBER_LIMIT setting such that disk space after cleanup
+    # allows another $n_scratch_prepost pre/post snapshot pairs
+    $number_limit_upper = $nmax_snapshots - $n_scratch_prepost * 2;
+    if ($number_limit_upper > 0) {
+        # plenty of free space? No ambition to fill it all, then.
+        $number_limit_upper = 18 if ($number_limit_upper > 18);
```

Comment:
where does the 18 come from?

Reply:
Each snapper_cleanup() iteration writes about $n_scratch_prepost * $scratchfile_mb to the disk: that is 3 GiB with the current settings (taken over from before my diff without change). Up to now the test encountered comparatively small filesystems (< 10 GB free) and did 4 (hardcoded) iterations. Now what should happen if, sometime in the future, the test finds, say, 30 GB to play with? Without a bound here it would set $number_limit_upper = ca. 52 and run "forever", to no real purpose. Introducing a bound is meant to forestall excessive runtime while still allowing the test to make its point. The value of 18 (resulting in 6 snapper_cleanup invocations) seemed a reasonable guess for me: it is open for debate, of course.